File: android/app/src/main/java/com/bkes994408/expensetracker/domain/AddTransactionUseCase.kt
package com.bkes994408.expensetracker.domain

class AddTransactionUseCase(
    private val repository: TransactionRepository,
) {
    suspend operator fun invoke(
        amountCents: Long,
        note: String,
        occurredAtEpochMillis: Long,
    ): Long {
        if (amountCents == 0L) {
            throw IllegalArgumentException("Amount must not be zero")
        }
        return repository.add(
            amountCents = amountCents,
            note = note.trim(),
            occurredAtEpochMillis = occurredAtEpochMillis,
        )
    }
}

File: ios/ExpenseTracker/Domain/AddTransactionUseCase.swift
import Foundation

enum AddTransactionValidationError: LocalizedError, Equatable {
    case zeroAmount

    var errorDescription: String? {
        switch self {
        case .zeroAmount:
            return "Amount must not be zero."
        }
    }
}

struct AddTransactionUseCase {
    let repository: TransactionRepository

    func callAsFunction(amountCents: Int64, note: String, occurredAt: Date) throws -> Int64 {
        guard amountCents != 0 else {
            throw AddTransactionValidationError.zeroAmount
        }
        return try repository.add(amountCents: amountCents, note: note.trimmingCharacters(in: .whitespacesAndNewlines), occurredAt: occurredAt)
    }
}
